### PR TITLE
Add custom exception when clusters are not found

### PR DESCRIPTION
**Pull request by andythsu**
## Description

If no adhoc clusters are found, we should throw a meaningful exception with some explanations.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
* 
```

### CODE REVIEW

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/QueryCountBasedRouter.java [hunk 1]
```diff
@@ -224,7 +224,7 @@ private synchronized Optional<String> getBackendForRoutingGroup(String routingGr
     @Override
     public String provideAdhocBackend(String user)
     {
-        return getBackendForRoutingGroup("adhoc", user).orElseThrow();
+        return getBackendForRoutingGroup("adhoc", user).orElseThrow(() -> new ClusterNotFoundException("did not find any cluster for the adhoc routing group"));
```
Comment (mosabua) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/QueryCountBasedRouter.java:227]: Should we make this more generic of an exception? Like RoutingException or so? Also should we pass in the magic `adhoc` from the prior code?

And update wording?

```
No cluster in routing group `adhoc` available.
```

Or similar

**Review by mosabua — CHANGES_REQUESTED** [review]
I like the idea of improving that experience .. but we need think about how to implement  that better

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/QueryCountBasedRouter.java (hunk 1 above)
Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/QueryCountBasedRouter.java:227]: we could put it the method `getBackendForRoutingGroup`. That method has the `routingGroup` passed in, so we can use that in our error message. That should be cleaner

Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/QueryCountBasedRouter.java:227]: Open to making the error more generic to be `RoutingException`

Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/QueryCountBasedRouter.java:227]: I think more general version, RoutingException is better :)

**Pull request closed without merging**